pyop2/codegen/builder.py

```diff
@@ -0,0 +1,816 @@
+            pack = GlobalPack(argument, arg.access)
+        elif arg._is_mat:
+            argument = Argument((), PetscMat(), pfx="mat")
+            map_ = tuple(self.map_(m) for m in arg.map)
+            pack = MatPack(argument, arg.access, map_,
+                           arg.data.dims, arg.data.dtype,
+                           interior_horizontal=interior_horizontal)
+        else:
+            raise ValueError("Unhandled argument type")
+        self.arguments.append(argument)
+        self.packed_args.append(pack)
+        self.argument_accesses.append(arg.access)
+
+    def map_(self, map_):
+        if map_ is None:
+            return None
+        interior_horizontal = self.iteration_region == ON_INTERIOR_FACETS
+        if isinstance(map_, DecoratedMap):
+            key = map_.map
+        else:
+            key = map_
+        try:
+            return self.maps[key]
+        except KeyError:
+            map_ = Map(map_, interior_horizontal,
+                       (self.bottom_layer, self.top_layer),
+                       self.indexed_variable_entity_masks)
+            self.maps[key] = map_
+            return map_
+
+    @property
+    def kernel_args(self):
+        return tuple(p.kernel_arg(self.loop_indices) for p in self.packed_args)
+
+    @property
+    def wrapper_args(self):
+        # Loop extents come from here.
+        args = list(self.forward_arguments)
+        args.extend(self._loop_index.extents)
+        if self.extruded:
+            args.append(self._layers_array)
+            if not self.constant_layers:
+                bottom, top = self.variable_entity_masks
+                assert bottom.offset == top.offset
+                args.append(bottom.offset)
+                args.append(bottom.values)
+                args.append(top.values)
+        if self.subset:
+            args.append(self._subset_indices)
+        # parloop args passed "as is"
+        args.extend(self.arguments)
+        # maps are refcounted
+        for map_ in self.maps.values():
+            args.append(map_.values)
+        return tuple(args)
+
+    def kernel_call(self):
+        args = self.kernel_args
+        access = tuple(self.argument_accesses)
+        import itertools
+        # assuming every index is free index
+        free_indices = set(itertools.chain.from_iterable(arg.multiindex for arg in args))
+        # remove runtime index
+        free_indices = tuple(i for i in free_indices if isinstance(i, Index))
+        if self.pass_layer_to_kernel:
+            args = args + (self.layer_index, )
+            access = access + (READ,)
+        if self.forward_arguments:
+            args = self.forward_arguments + args
+            access = tuple([WRITE] * len(self.forward_arguments)) + access
+        return FunctionCall(self.kernel.name, KernelInst(), access, free_indices, *args)
+
+    def emit_instructions(self):
+        yield DummyInstruction(PackInst(), *self.loop_indices)
```

Comment:
I think I need at least one PackInst to anchor the kernel call instructions. Otherwise, I think the problem is the kernel call is not nested correctly (IIRC).